core/api/transport/listener.hpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_CORE_API_TRANSPORT_LISTENER_HPP
+#define KAGOME_CORE_API_TRANSPORT_LISTENER_HPP
+
+#include <memory>
+#include <unordered_map>
+
+#include <boost/asio/io_context.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include "api/transport/session.hpp"
+#include <outcome/outcome.hpp>
+
+namespace kagome::server {
+  /**
+   * @brief server which listens for incoming connection,
+   * accepts connections making session from socket
+   */
+  class Listener {
+   protected:
+    template <class T>
+    using sptr = std::shared_ptr<Session>;
+
+    template <class T>
+    using Signal = boost::signals2::signal<T>;
+    using OnNewSession = Signal<void(sptr<Session>)>;
```

Comment:
hm. do you have many subscribers "on new session" or just one?

Reply:
just one `service` subscriber: many sessions - one service.

Reply:
If u have one subscriber, it is easier to use plain call back instead of signal

Reply:
In this case I will have to inverse initialization, for example transfer a slot to listener constructor. Not sure if it is a good idea. Currently, listener doesn't know about `consumer`, I wouldn't want to change that.

Reply:
Transport can have a method 
`start(callback)`, which will be executed by client. e.g. it starts read/write loop and sets processing callback at the same time

Reply:
done